In [658]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import SGDRegressor
import seaborn as sns
from sklearn.pipeline import Pipeline
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV
from sklearn.tree import ExtraTreeRegressor
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.feature_selection import VarianceThreshold
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import mutual_info_regression

# Assignment 2 - Regression
### Predict the TOTAL COMPENSATION for this year. 

The data file provided is a salary survey for tech workers in Europe. We want to predict the total amount of compensation they bring in each year, based off of the details of their work. 

Some notes that will be important:
<ul>
    <li>The total compensation will need to be constructed, there is a column for salary, "Yearly brutto salary (without bonus and stocks) in EUR", as well as a column for bonus compensation, "Yearly bonus + stocks in EUR". 
    <li>Some categorical variables will need some work, and there isn't generally an exact answer. The main concern is things with categories that have a bunch of values with a very small count. For example, if there is only 1 person in City X, then that value likely needs to be addressed. We don't want it encoded into a new column of one 1 and thousands of 0s. 
    <li>There is an article exploring some of the data here: https://www.asdcode.de/2021/01/it-salary-survey-december-2020.html
    <li>Imputation and a bit of data manipulation will be required. 
    <li>Use any regression method you'd like. Some ones are closely related to what we've done, you may want to look at them, e.g. ExtraTreesRegressor. 
    <li>Initial accurracy, and potentially final accuracy, may not be great. When I made a plain model will little optimization the errors were large and the R2 was low. There is lots of room for optimization. 
    <li>Research challenge - try some work on the target, look into TransformedTargetRegressor and see if that helps. Recall in stats when we had skewed distributions... Maybe it helps, maybe it doesn't. 
    <li>EDA and data prep are up to you - you'll probably need to do a little exploring to figure out what cleanup is needed. When I did it, I did things kind of iteratively when I did it. For example, look at the value counts, figure out how to treat the different categories, clean something up, look at the results, potentially repeat if needed. After you figure out what needs to be done, you may be able to take some of those steps and incorporate them into a pipeline to be cleaner....
    <li><b>CRITICAL - Please make sure your code runs with RUN ALL. It should load the data that you're given, do all the processing, and spit out results. Comment out or remove anything that you've cleaned up and don't need - e.g. if you scaled a value manually, then moved that into a pipeline, don't leave the original scaling code active when the file is run.</b>
</ul>

### Details and Deliverables

You'll need to build code to produce the predictions. In particular, there's a few things that'll be marked:
<ul>
    <li>Please add a markdown cell at the bottom, and put in a few notes addressing the following:
    <ul>
        <li> Accuracy of your models with/without feature selection. Include both train/test for each. Please use R2 and RMSE. 
        <li> Feature Selection - Please identify what you did for feature selection. No need for a long explaination, something along the lines of "I did X, and the result was that 4 features were removed". Try at least 2 things. 
        <li> Hyperparameter Changes / Grid Search Improvements. What did you try, and why. Similar explaination to above, short. 
        <li> Overall this section should be roughly as long as this intro block - just outline what the results were, what you did to improve, and the results after. 
        <li> If you could use titles/bullet points I'd really appreciate it. 
    </ul>
    <li>Grade Breakdown:
    <ul>
        <li> Code is readable, there are comments: 20%
        <li> Explaination as defined above: 60% (20% each point)
        <li> Accuracy: 20% As compared to everyone else. This will be generously graded, I won't be surprised if overall accuracy is low for most people. 
    </ul>
</ul>

In [659]:
#Load Data
df = pd.read_csv("data/Euro_Salary.csv")
df.describe(include="all").T

count unique  \
Timestamp                                             1253   1248   
Age                                                 1226.0    NaN   
Gender                                                1243      3   
City                                                  1253    119   
Position                                              1247    148   
Total years of experience                             1237     48   
Years of experience in Germany                        1221     53   
Seniority level                                       1241     24   
Your main technology / programming language           1126    256   
Other technologies/programming languages you us...    1096    562   
Yearly brutto salary (without bonus and stocks)...  1253.0    NaN   
Yearly bonus + stocks in EUR                           829    168   
Number of vacation days                               1185     45   
Employment status                                     1236     11   
Сontract duration                                     1224      3   
Main language at work                                 1237     14   
Company size                                          1235      5   
Company type                                          1228     63   

                                                                        top  \
Timestamp                                               24/11/2020 13:55:19   
Age                                                                     NaN   
Gender                                                                 Male   
City                                                                 Berlin   
Position                                                  Software Engineer   
Total years of experience                                                10   
Years of experience in Germany                                            2   
Seniority level                                                      Senior   
Your main technology / programming language                            Java   
Other technologies/programming languages you us...  Javascript / Typescript   
Yearly brutto salary (without bonus and stocks)...                      NaN   
Yearly bonus + stocks in EUR                                              0   
Number of vacation days                                                  30   
Employment status                                        Full-time employee   
Сontract duration                                        Unlimited contract   
Main language at work                                               English   
Company size                                                          1000+   
Company type                                                        Product   

                                                    freq             mean  \
Timestamp                                              2              NaN   
Age                                                  NaN        32.509788   
Gender                                              1049              NaN   
City                                                 681              NaN   
Position                                             387              NaN   
Total years of experience                            138              NaN   
Years of experience in Germany                       195              NaN   
Seniority level                                      565              NaN   
Your main technology / programming language          184              NaN   
Other technologies/programming languages you us...    44              NaN   
Yearly brutto salary (without bonus and stocks)...   NaN  80279042.578715   
Yearly bonus + stocks in EUR                         227              NaN   
Number of vacation days                              488              NaN   
Employment status                                   1190              NaN   
Сontract duration                                   1159              NaN   
Main la

In [660]:
df.sample(10)

Timestamp   Age  Gender       City             Position   \
960   30/11/2020 13:33:21  28.0    Male  Frankfurt          XR Developer   
793   26/11/2020 21:13:28  32.0    Male     Berlin  Senior Data Engineer   
497   25/11/2020 09:05:26  38.0    Male     Berlin        Data Scientist   
472   25/11/2020 01:29:48  31.0  Female     Berlin           QA Engineer   
635   25/11/2020 16:16:23  33.0    Male     Berlin                DevOps   
208   24/11/2020 13:07:24  38.0    Male     Berlin     Software Engineer   
1044  01/12/2020 21:28:19  34.0    Male    Hamburg                DevOps   
1178  19/12/2020 09:31:16  31.0     NaN     Berlin           ML Engineer   
33    24/11/2020 11:27:49  39.0    Male     Berlin     Software Engineer   
63    24/11/2020 11:37:52  34.0    Male     Munich     Software Engineer   

     Total years of experience Years of experience in Germany Seniority level  \
960                          8                              1            Lead   
793                          9                              4          Senior   
497                         15                              2            Head   
472                         11                            1.5          Middle   
635                          8                              3          Senior   
208                         16                              1          Senior   
1044                         8                              1          Senior   
1178                         5                            2.5          Middle   
33                           8                              5          Senior   
63                          13                              3            Lead   

     Your main technology / programming language  \
960                                           C#   
793                                       Python   
497                                       Python   
472                                           JS   
635                                          AWS   
208                                       Python   
1044                                    Embedded   
1178                                      Python   
33                                          Java   
63                                          .NET   

     Other technologies/programming languages you use often  \
960                     Python, Javascript / Typescript       
793                              Java / Scala, SQL, AWS       
497                        Python, SQL, SAP / ABAP, AWS       
472                                           Ruby, SQL       
635                     Python, AWS, Kubernetes, Docker       
208   Python, Javascript / Typescript, Java / Scala,...       
1044                                              C/C++       
1178  Java / Scala, AWS, Google Cloud, Kubernetes, D...       
33    Python, Java / Scala, SQL, AWS, Google Cloud, ...       
63                                            .NET, SQL       

      Yearly brutto salary (without bonus and stocks) in EUR  \
960                                             25000.0        
793                                             90000.0        
497                                            100000.0        
472                                             57000.0        
635                                             75000.0        
208                                             65000.0        
1044                                            67500.0        
1178                                            77500.0        
33                                              80000.0        
63                                              72000.0        

     Yearly bonus + stocks in EUR Number of vacation days   Employment status  \
960                          2000                       1  Full-time employee   
793                           NaN                      30  Full-time employee   
497                        120000                      27  Full-time emp

In [661]:
#df.info()
df["Other technologies/programming languages you use often"].isna().value_counts()

False    1096
True      157
Name: Other technologies/programming languages you use often, dtype: int64

In [662]:
df.head()

Timestamp   Age Gender    City           Position   \
0  24/11/2020 11:14:15  26.0   Male  Munich   Software Engineer   
1  24/11/2020 11:14:16  26.0   Male  Berlin   Backend Developer   
2  24/11/2020 11:14:21  29.0   Male  Berlin   Software Engineer   
3  24/11/2020 11:15:24  28.0   Male  Berlin  Frontend Developer   
4  24/11/2020 11:15:46  37.0   Male  Berlin   Backend Developer   

  Total years of experience Years of experience in Germany Seniority level  \
0                         5                              3          Senior   
1                         7                              4          Senior   
2                        12                              6            Lead   
3                         4                              1          Junior   
4                        17                              6          Senior   

  Your main technology / programming language  \
0                                  TypeScript   
1                                        Ruby   
2                     Javascript / Typescript   
3                                  Javascript   
4                                     C# .NET   

  Other technologies/programming languages you use often  \
0                    Kotlin, Javascript / Typescript       
1                                                NaN       
2                    Javascript / Typescript, Docker       
3                                                NaN       
4                             .NET, SQL, AWS, Docker       

   Yearly brutto salary (without bonus and stocks) in EUR  \
0                                            80000.0        
1                                            80000.0        
2                                           120000.0        
3                                            54000.0        
4                                            62000.0        

  Yearly bonus + stocks in EUR Number of vacation days  \
0                         5000                      30   
1                          NaN                      28   
2                       120000                      30   
3                          NaN                      24   
4                          NaN                      29   

            Employment status   Сontract duration Main language at work  \
0          Full-time employee  Unlimited contract               English   
1          Full-time employee  Unlimited contract               English   
2  Self-employed (freelancer)  Temporary contract               English   
3          Full-time employee  Unlimited contract               English   
4          Full-time employee  Unlimited contract               English   

  Company size Company type  
0       51-100      Product  
1     101-1000      Product  
2     101-1000      Product  
3       51-100      Startup  
4     101-1000      Product

In [663]:
#create the target column of Total compensation by combining the yearly salary and yearly bonus columns, then dropping them as well as a metadata column

df["Yearly bonus + stocks in EUR"] = pd.to_numeric(df["Yearly bonus + stocks in EUR"], downcast="float", errors="coerce")
df["Yearly bonus + stocks in EUR"].fillna(0, inplace=True)
df['Total Compensation'] = df['Yearly brutto salary (without bonus and stocks) in EUR'] + df['Yearly bonus + stocks in EUR']

df.drop(columns={"Timestamp","Yearly brutto salary (without bonus and stocks) in EUR", "Yearly bonus + stocks in EUR"}, inplace=True)
df['Position '].rename('Position')

0                            Software Engineer
1                            Backend Developer
2                            Software Engineer
3                           Frontend Developer
4                            Backend Developer
                         ...                  
1248                         Backend Developer
1249    Researcher/ Consumer Insights Analyst 
1250                     IT Operations Manager
1251                        Frontend Developer
1252                                    DevOps
Name: Position, Length: 1253, dtype: object

In [664]:
#function to replace the x lowest occurring values in a column with a designated value
def replace_low_freq(d, col, threshold=10, replacement='other'):
    value_counts = d[col].value_counts() # Specific column 
    to_remove = value_counts[value_counts <= threshold].index
    tmp = d[col].replace(to_replace=to_remove, value=replacement)
    return tmp

In [665]:
#removing the excessive amounts of long responses in these two columns
df["Your main technology / programming language"] = replace_low_freq(df, "Your main technology / programming language", 10, "Other")
df['Other technologies/programming languages you use often'] = replace_low_freq(df, "Your main technology / programming language", 5, "Other")

In [666]:
#Filling in NaNs with selected data. For categorical columns, they are labeled as unknown. For numerical columns, they are usually filled with the median column value
df['Age'].fillna(df['Age'].median(), inplace=True)

df['Gender'].fillna('Other',inplace=True)
df['Position '].fillna('Other',inplace=True)
df['Seniority level'].fillna('Other',inplace=True)
df['Company type'].fillna('Other',inplace=True)

df['Employment status'].fillna('Unknown',inplace=True)
df['Сontract duration'].fillna('Unknown',inplace=True)
df['Main language at work'].fillna('Unknown',inplace=True)
df['Company size'].fillna('Unknown',inplace=True)
df['Company type'].fillna('Unknown',inplace=True)

df['Years of experience in Germany'] = pd.to_numeric(df['Years of experience in Germany'], downcast="float", errors="coerce")
df['Years of experience in Germany'].fillna(df['Years of experience in Germany'].median(),inplace=True)

df['Number of vacation days'] = pd.to_numeric(df['Number of vacation days'], downcast="float", errors="coerce")
df['Number of vacation days'].fillna(df['Number of vacation days'].median(),inplace=True)

df['Total years of experience'] = pd.to_numeric(df['Total years of experience'], downcast="float", errors="coerce")
df['Total years of experience'].fillna(df['Years of experience in Germany'],inplace=True)

df["Your main technology / programming language"].fillna('Unknown',inplace=True)
df['Other technologies/programming languages you use often'].fillna('Unknown',inplace=True)

In [667]:
#creation of the full feature pipeline using a Column Transformer
categorical_features = ['Gender', 'City', 'Position ','Seniority level','Your main technology / programming language',
                        'Other technologies/programming languages you use often','Employment status','Сontract duration',
                        'Main language at work','Company size', 'Company type']
categorical_transformer = OneHotEncoder(handle_unknown = 'ignore')

numeric_features = ["Age", 'Total years of experience','Years of experience in Germany','Number of vacation days']
numeric_transformer = Pipeline( steps=[
        ("imputer", SimpleImputer(strategy="mean")),
        ("scaler", MinMaxScaler())
    ])

preprocessor = ColumnTransformer( transformers=[
        ("num", numeric_transformer, numeric_features),
        ("cat", categorical_transformer, categorical_features),
    ])

rf_para = {"criterion":['squared_error','friedman_mse', 'absolute_error', 'poisson'],
            'max_depth':[1,2,3,4,5],
            'ccp_alpha':[0.020,0.0205,0.021],
            'splitter':['random','best'],
            'max_features':['None', 'sqrt', 'log2']
            }

#Creating the models, with a Grid Search used on the Extra Tree Regressor
pipeline_steps_grid = [('pre', preprocessor),('search', GridSearchCV(estimator=ExtraTreeRegressor(),n_jobs=-1,param_grid=rf_para))]
grid = Pipeline(pipeline_steps_grid)

In [668]:
y = (df['Total Compensation'])
x = df.drop(columns={'Total Compensation'})

xTrain,xTest,yTrain,yTest = train_test_split(x,y,random_state=0)

In [669]:
#Fitting and output of RMSE and R2 values of each model
grid.fit(xTrain,yTrain)

yPredG = grid.predict(xTest)

print(grid.named_steps['search'].best_estimator_)
print("RMSE:",mean_squared_error(yTest,yPredG,squared=False),'\n',"R2 Train:",grid.score(xTrain,yTrain),'\n',"R2 Test:",grid.score(xTest,yTest))

ExtraTreeRegressor(ccp_alpha=0.021, criterion='absolute_error', max_depth=4,
                   max_features='sqrt', splitter='best')
RMSE: 50985.039471707074 
 R2 Train: -0.0011818614362539837 
 R2 Test: -0.02799227738493726


c:\Users\atrsy\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:378: FitFailedWarning: 
600 fits failed out of a total of 1800.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
600 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\atrsy\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\atrsy\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 1342, in fit
    super().fit(
  File "c:\Users\atrsy\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 317, in fit
    raise ValueError(
ValueError: Invalid value for max_features. Allowed string values are 'auto'

In [670]:
#Feature selection using Variance Threshold
varTh = VarianceThreshold(.2)
xDummied = pd.get_dummies(x,drop_first=True)
fsX = varTh.fit_transform(xDummied,y)
mask = varTh.get_support()
newFeatures = xDummied.columns[mask]
print(newFeatures)

Index(['Age', 'Total years of experience', 'Years of experience in Germany',
       'Number of vacation days', 'City_Berlin', 'Position _Software Engineer',
       'Seniority level_Middle', 'Seniority level_Senior',
       'Your main technology / programming language_Other',
       'Other technologies/programming languages you use often_Other',
       'Company size_101-1000', 'Company type_Product'],
      dtype='object')


In [671]:
kb = SelectKBest(mutual_info_regression, k=7)
kbX = kb.fit_transform(xDummied,y)
print("K-Best:", kbX.shape)

#Print the features
mask = kb.get_support()
new_features = xDummied.columns[mask]
print("K-Best Features:", new_features)

K-Best: (1253, 7)
K-Best Features: Index(['Age', 'Total years of experience', 'Years of experience in Germany',
       'Number of vacation days', 'Seniority level_Middle',
       'Seniority level_Senior', 'Сontract duration_Unlimited contract'],
      dtype='object')


In [672]:
#creation of the feature selection pipeline.
numericTransformer = Pipeline(steps=[
        ("imputer", SimpleImputer(strategy="mean")),
        ("scaler", MinMaxScaler())
    ])

##Creating the models, with a Grid Search used on the Extra Tree Regressor
pipelineStepsGrid = [('num', numericTransformer),('search', GridSearchCV(estimator=ExtraTreeRegressor(),n_jobs=-1,param_grid=rf_para))]
fsGrid = Pipeline(pipelineStepsGrid)

In [674]:
#Fitting and output of RMSE and R2 values of each model
fsX = pd.DataFrame(fsX,columns=newFeatures)
xTrainFS,xTestFS,yTrainFS,yTestFS = train_test_split(fsX, y,random_state=42)

fsGrid.fit(xTrainFS,yTrainFS)
yPredFSG = fsGrid.predict(xTestFS)

print("Variance Threshold",fsGrid.named_steps['search'].best_estimator_)
print('RMSE:',mean_squared_error(yTestFS,yPredFSG,squared=False),'\n',"R2 Train:",fsGrid.score(xTrainFS,yTrainFS),'\n',"R2 Test:",fsGrid.score(xTestFS,yTestFS),'\n')

kbX = pd.DataFrame(kbX,columns=new_features)
xTrainFS,xTestFS,yTrainFS,yTestFS = train_test_split(kbX, y,random_state=42)

fsGrid.fit(xTrainFS,yTrainFS)
yPredFSG = fsGrid.predict(xTestFS)

print("Select K Best",fsGrid.named_steps['search'].best_estimator_)
print('RMSE:',mean_squared_error(yTestFS,yPredFSG,squared=False),'\n',"R2 Train:",fsGrid.score(xTrainFS,yTrainFS),'\n',"R2 Test:",fsGrid.score(xTestFS,yTestFS))

c:\Users\atrsy\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:378: FitFailedWarning: 
600 fits failed out of a total of 1800.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
600 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\atrsy\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\atrsy\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 1342, in fit
    super().fit(
  File "c:\Users\atrsy\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 317, in fit
    raise ValueError(
ValueError: Invalid value for max_features. Allowed string values are 'auto'

Variance Threshold ExtraTreeRegressor(ccp_alpha=0.021, criterion='absolute_error', max_depth=4,
                   max_features='sqrt', splitter='best')
RMSE: 70665.15203175523 
 R2 Train: 0.0018359457258604328 
 R2 Test: -0.0042290345237328 

Select K Best ExtraTreeRegressor(ccp_alpha=0.0205, criterion='absolute_error', max_depth=5,
                   max_features='log2', splitter='best')
RMSE: 67007.47184415333 
 R2 Train: 0.0018354666232668349 
 R2 Test: 0.09703972678997608


c:\Users\atrsy\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:378: FitFailedWarning: 
600 fits failed out of a total of 1800.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
600 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\atrsy\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "c:\Users\atrsy\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 1342, in fit
    super().fit(
  File "c:\Users\atrsy\anaconda3\lib\site-packages\sklearn\tree\_classes.py", line 317, in fit
    raise ValueError(
ValueError: Invalid value for max_features. Allowed string values are 'auto'

# Answers and Explainations
(Expand/modify as needed)

### Results
<ul>
    <li>The best model seemed to be the Extra Tree Regressor, having an RMSE of 50971.06, a Training R2 of -0.00118, and a Testing R2 of -0.0274
    <li>After feature selection, best RMSE was 67007.472, Training R2 was 0.00184 and Testing R2 was 0.0970
</ul>

### Feature Selection Activities
<ul>
    <li>Variance Threshold, removed 4 features
    <li>Select K Best, kept 7 features
</ul>  

### Hyperparameter Changes
<ul>
    <li>Criterion - squared_error, friedman_mse, absolute_error, poisson
    <li>Max_depth - 1, 2, 3, 4, 5
    <li>ccp_alpha - 0.020, 0.0205, 0.021
    <li>Splitter - random, best
    <li>Max_features - None, sqrt, log2
</ul>